src/com/unbounce/treajure/concurrent/mvar.clj

```diff
@@ -0,0 +1,82 @@
+(ns com.unbounce.treajure.concurrent.mvar
+  "Haskell's MVar abstraction in Clojure.
+
+  Check: http://chimera.labs.oreilly.com/books/1230000000929/ch07.html for more
+  info around MVars."
+  (:import [java.util.concurrent Semaphore]))
+
+;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
+;; Lightweight Performance mutable tuple implementation in Clojure
+;; (Look ma! no atoms!)
+
+(defprotocol IMCons
+  (mcar [p])
+  (mcdr [p])
+  (set-mcar! [p val])
+  (set-mcdr! [p val]))
+
+(deftype MCons [^{:volatile-mutable true} car
+                ^{:volatile-mutable true} cdr]
```

Comment:
I think it is worth mentioning that the reason why I'm adding this abstraction is to implement a sync strategy between Observables from Hystrix and requirements for ring/jetty server sync implementations.

An example of the [usage](https://github.com/unbounce/lp-auth/blob/lp-7803-auth-spike/src/lp_auth/util/rx.clj#L28) can be found on [lp-auth's demo](https://github.com/unbounce/lp-auth/blob/lp-7803-auth-spike/src/lp_auth/core.clj#L151)